homeassistant/components/viam/__init__.py

```diff
@@ -0,0 +1,367 @@
+        confidence_threshold = float(call.data.get(SERVICE_CONFIDENCE, 0.6))
+
+        async with await self._get_robot_client(
+            call.data.get(SERVICE_ROBOT_SECRET), call.data.get(SERVICE_ROBOT_ADDRESS)
+        ) as robot:
+            detector = VisionClient.from_robot(robot, detector_name)
+            image = await self._get_image(filepath, camera)
+
+        if image is None:
+            return {
+                "detections": [],
+                "img_src": filepath or None,
+            }
+
+        img_src = filepath or self._encode_image(image)
+        detections = await detector.get_detections(image)
+
+        return {
+            "detections": [
+                {
+                    "name": c.class_name,
+                    "confidence": c.confidence,
+                    "x_min": c.x_min,
+                    "y_min": c.y_min,
+                    "x_max": c.x_max,
+                    "y_max": c.y_max,
+                }
+                for c in detections
+                if c.confidence >= confidence_threshold
+            ],
+            "img_src": img_src,
+        }
+
+    def _encode_image(self, image: Image.Image | RawImage):
+        image_bytes = b""
+        if isinstance(image, Image.Image):
+            image_bytes = image.tobytes()
+        if isinstance(image, RawImage):
+            image_bytes = image.data
+
+        image_string = (
+            str(base64.b64encode(image_bytes))
+            .replace("\\n", "")
+            .replace("b'", "")
+            .replace("'", "")
+        )
+        return f"data:image/jpeg;base64,{image_string}"
+
+    async def _get_image(self, filepath: str | None, camera: str | None):
+        if filepath is not None:
+            return await self.hass.async_add_executor_job(_fetch_image, filepath)
+        if camera is not None:
+            cam_bytes = await self._get_image_from_camera(camera)
+            if len(cam_bytes) > 0:
+                return RawImage(cam_bytes, "jpeg")
+
+        return None
+
+    async def _get_image_from_camera(self, camera_name: str) -> bytes:
+        cam_entity = er.async_get(self.hass).async_get(camera_name)
+        if cam_entity is None:
+            raise ServiceValidationError(
+                f"A camera entity called {camera_name} was not found",
+                translation_domain=DOMAIN,
+                translation_key="camera_entity_not_found",
+                translation_placeholders={
+                    "camera_name": camera_name,
+                },
+            )
+
+        cam = self.hass.data[cam_entity.domain].get_entity(camera_name)
+
+        if cam is None:
+            raise ServiceValidationError(
+                f"A camera for the entity {camera_name} was not found",
+                translation_domain=DOMAIN,
+                translation_key="camera_not_found",
+                translation_placeholders={"camera_name": camera_name},
+            )
+
+        return await cam.async_camera_image()
```

Comment:
Accessing the camera entity directly isn't allowed. Please use `async_get_image` from the camera integration instead.